### ISSUE TITLE
`SchemaError` when using `Field` length constraints with list of objects derived from `BaseModel`

### ISSUE THREAD

**Issue by rodolphebarbanneau** [pydantic/pydantic#5542]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

Hello Pydantic!

I've encountered an issue with the [v2.0a2 2023-04-12](https://github.com/pydantic/pydantic/releases/tag/v2.0a2) release where an error is raised when a constraint is applied to a subsequent list of the same type derived from `BaseModel`.

Below are two test cases to demonstrate the issue with some `list[Movie]` annotated attributes where `Movie` inherits from `BaseModel` such that:
```python
class Movie(BaseModel):
    name: str
```

### ✅ Case 1
When applying a length constraint (e.g., `min_length=1`) to the first `list[Movie]`, there is no problem.
```python
class Model(BaseModel):
    liked: list[Movie] = Field(min_length=1)
    disliked: list[Movie]

model = Model(liked=[Movie(name='John Wick')], disliked=[Movie(name='Exorcist II')]) # OK
```

### ❌ Case 2
When applying a length constraint (e.g., `min_length=1`) to a subsequent `list[Movie]`, it breaks...
```python
class Model(BaseModel):
    liked: list[Movie]
    disliked: list[Movie] = Field(min_length=1)

model = Model(liked=[Movie(name='John Wick')], disliked=[Movie(name='Exorcist II')]) # Error
```
```
Error: pydantic_core._pydantic_core.SchemaError: Error building "list" validator:
    SchemaError: Error building "definition-ref" validator:
    SchemaError: Slots Error: ref '__main__.Movie:140226593046736' not found
```

It seems that the issue arises only when a constraint is applied to a list of the same type derived from `BaseModel` after the first occurrence of that list type.

### Example Code

```Python
from pydantic import BaseModel, Field


#class Movie:
#    def __init__(self, name:str):
#        self.name = name

class Movie(BaseModel):
    name: str


"""
TEST 1: OK
"""
class Model(BaseModel, arbitrary_types_allowed=True):
    liked: list[Movie] = Field(min_length=1)
    disliked: list[Movie]

model = Model(liked=[Movie(name='John Wick')], disliked=[Movie(name='Exorcist II')]) # OK


"""
TEST 2: KO
Error: pydantic_core._pydantic_core.SchemaError: Error building "list" validator:
    SchemaError: Error building "definition-ref" validator:
    SchemaError: Slots Error: ref '__main__.Movie:140226593046736' not found
"""
class Model(BaseModel, arbitrary_types_allowed=True):
    liked: list[Movie]
    disliked: list[Movie] = Field(min_length=1)

model = Model(liked=[Movie(name='John Wick')], disliked=[Movie(name='Exorcist II')]) # Error
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0a2
pydantic-core version: 0.23.1 release build profile
install path: [...]/pypoetry/virtualenvs/tests-UcNkclcz-py3.11/lib/python3.11/site-packages/pydantic
python version: 3.11.0 (v3.11.0:deaf509e8f, Oct 24 2022, 14:43:23) [Clang 13.0.0 (clang-1300.0.29.30)]
platform: macOS-13.3.1-x86_64-i386-64bit
optional deps. installed: ['typing-extensions']
```

**Comment by rodolphebarbanneau** [comment]
I've just tried the same code with a3 and core 0.25 and it seems you have already corrected this with the new release.

**Comment by hramezani** [comment]
Thanks @rodolphebarbanneau for testing and reporting the bug. We highly appreciate it. Please help us to improve Pydantic by testing.
I can confirm the problem was there in `v2.0a2` and got fixed in `v2.0a3` probably in https://github.com/pydantic/pydantic/commit/7e9a5489c3445267788e7667342f6ae446d45595.
